validate_argument_spec - fix validating the argument spec before using it to validate vars

s-hertel opened the pull request:
##### SUMMARY
add a spec for the spec

This resolves the error in #79624, but replaces it with a warning because `no_log` in the role's argument spec has no bearing on whether or not the tasks in the role divulge it. The role currently needs to manage things like `no_log`/`aliases` for variables itself.

Fixes #79624

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
validate_argument_spec

s-hertel commented on lib/ansible/playbook/role/__init__.py:
```diff
@@ -262,11 +262,18 @@ def _load_role_data(self, role_include, parent_role=None):
+            )
+            for block in validation_blocks:
+                for task in block.block:
+                    task.implicit = True
```
This won't change how the validation task is displayed by default. Callbacks only ignore implicit Tasks (unless they set the attr `wants_implicit_tasks`), and always get the TaskResults, even from implicit Tasks. https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_queue_manager.py#L442-L443

s-hertel commented on lib/ansible/plugins/action/validate_argument_spec.py:
```diff
@@ -77,6 +109,44 @@ def run(self, tmp=None, task_vars=None):
+                    options=dict(type='dict'),
+                    required=dict(type='bool'),
+                    # don't deter documentation, the argparser will ignore these
+                    version_added=dict(),
+                    version_added_collection=dict(),
```
Wondering what others would think about including these two. It was so flexible before, that I worry the warnings might be noisy/unhelpful in some cases.

felixfontein replied:
If this is only about role argument specs, I don't think `version_added_collection` makes any sense, since options cannot come from other collections. (There are no doc fragments for role argument specs.)

`version_added` has been supported by antsibull-docs since the beginning.

s-hertel replied:
It would only have occurred to me to use `version_added` for standalone roles... are you using collection versions in the `version_added` field?

Oh - I was remember the assortment of ways one can specify the removal/deprecation version/date and thinking these were two options accomplishing basically the same thing, and didn't remember the second is the collection name. I agree, only `version_added` makes sense here. Thanks for weighing in, this is now resolved from my perspective.

samccann reviewed (approved):
docs portion LGTM

felixfontein commented on lib/ansible/plugins/action/validate_argument_spec.py:
```diff
@@ -77,6 +109,44 @@ def run(self, tmp=None, task_vars=None):
+            option_types = [None, 'bits', 'bool', 'bytes', 'dict', 'float', 'int', 'json', 'jsonarg', 'list', 'path', 'raw', 'sid', 'str']
+            meta_validator = ArgumentSpecValidator(
+                dict(
+                    description=dict(),
```
Does `=dict()` mean "this can be any type", or does it mean "optional string"? If the latter, this isn't correct, since descriptions can also be lists of strings.

s-hertel replied:
Thanks, good catch. This was just a placeholder value really, since the key is expected but not necessary to validate for option validation (and wouldn't fail the type conversion), but I updated this to a list of str elements. And for the same reason, updated `default` to type raw.

felixfontein commented on changelogs/fragments/79657-warn-for-unsupported-role-arg-spec-fields.yml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - roles - give a warning for unknown fields during implicit M(validate_argument_spec) tasks.
```
```suggestion
  - roles - give a warning for unknown fields during implicit ``validate_argument_spec`` tasks.
```
`M(...)` only works in documentation, not in changelogs. (And even if it would, you'd need `M(ansible.builtin.validate_argument_spec)` here.)

s-hertel committed:
Fix validate_argument_spec so the argument spec is validated before
passing it to the ArgumentSpecValidator to prevent unhandled errors
during variable validation.

Make role argument spec errors non-fatal at runtime. Display a warning
and include details with -vvv. Pass ArgumentSpecValidator a pruned
argument spec containing the valid portion.

Give an error and return argument_spec_errors in the task result for
non-implicit validate_argument_spec tasks.

Add tests for role argument spec errors and validate_argument_spec
argument_spec errors.

s-hertel commented on lib/ansible/plugins/action/validate_argument_spec.py:
```diff
@@ -6,8 +6,111 @@
+            fields.pop(field, None)
+
+    if fields.get('type') in ('list', 'dict') and fields.get('options'):
+        validate_argspec(validator, context, fields['options'], errors, prune)
```
Note: suboptions can use `required_if`, `mutually_exclusive`, etc, so not supporting those for this code path would be a breaking change. Maybe this needs a deprecation period just to be safe. It would be simple just to allow those fields for top level options (instead of only suboptions) to fix the inconsistency, but [the plan](https://github.com/ansible/ansible/issues/74995#issuecomment-860931188) was to implement a better UX.
